### PR TITLE
AArch64: Implement ArrayCHK Evaluator

### PR DESCRIPTION
**Pull request by aarongraham9**
Implements ArrayCHK Evaluator for AArch64

Co-authored-by: Aaron Graham <aaron.graham@unb.ca>
Co-authored-by: Md. Mahbubur Rahman <mrahma15@unb.ca>

Signed-off-by: Aaron Graham <aaron.graham@unb.ca>

### CODE REVIEW

**Comment by aarongraham9** [comment]
This PR: subsumes and continues PR #6297.

Rebased off of the `master` branch.

This should be good for a review @0xdaryl and @knn-k.

**Comment by aarongraham9** [comment]
This needs a rebase, but otherwise, should be good for a review @0xdaryl.

**Comment by aarongraham9** [comment]
Rebased off of `master` and fixed up the merge conflicts.

**Comment by knn-k** [comment]
This PR changes the permission of the files from 100644 to 100755.  Please keep the original permission unless you need to change it.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp [hunk 1]
```diff
@@ -322,8 +322,174 @@ J9::ARM64::TreeEvaluator::DIVCHKEvaluator(TR::Node *node, TR::CodeGenerator *cg)
...
-   cg->evaluate(node->getFirstChild());
+TR::Register *
+J9::ARM64::TreeEvaluator::ArrayCHKEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   TR::Register *obj1Reg = cg->evaluate(node->getFirstChild());
+   TR::Register *obj2Reg = cg->evaluate(node->getSecondChild());
+   TR::Register *tmp1Reg = cg->allocateRegister();
+   TR::Register *tmp2Reg = cg->allocateRegister();
+   TR::LabelSymbol *doneLabel = generateLabelSymbol(cg);
+   TR::RegisterDependencyConditions *deps = new (cg->trHeapMemory()) TR::RegisterDependencyConditions(2, 2, cg->trMemory());
```
Comment (knn-k) [inline runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp]: The size of the allocated `RegisterDependencyConditions` is wrong.  You call `TR::addDependency()` on 4 registers below.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.hpp [hunk 2]
```diff
@@ -53,6 +53,8 @@ class OMR_EXTENSIBLE TreeEvaluator: public J9::TreeEvaluator
 
    static TR::Register *DIVCHKEvaluator(TR::Node *node, TR::CodeGenerator *cg);
 
+   static TR::Register *arrayCHKEvaluator(TR::Node *node, TR::CodeGenerator *cg);
```
Comment (knn-k) [inline runtime/compiler/aarch64/codegen/J9TreeEvaluator.hpp]: The name of the evaluator is inconsistent with the .cpp file.  (`arrayCHK` or `ArrayCHK`).

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp [hunk 3]
```diff
@@ -322,8 +322,174 @@ J9::ARM64::TreeEvaluator::DIVCHKEvaluator(TR::Node *node, TR::CodeGenerator *cg)
...
+   if ((!node->isArrayChkPrimitiveArray1())
+         && (!node->isArrayChkReferenceArray1())
+         && (!node->isArrayChkPrimitiveArray2())
+         && (!node->isArrayChkReferenceArray2()))
+      {
+#ifdef OMR_GC_COMPRESSED_POINTERS
+      generateTrg1MemInstruction(cg, TR::InstOpCode::ldrw, node, tmp1Reg,
+            new (cg->trHeapMemory()) TR::MemoryReference(obj1Reg, (int32_t) TR::Compiler->om.offsetOfObjectVftField(), cg));
+#else
+      generateTrg1MemInstruction(cg, isLong ? TR::InstOpCode::ldrx : TR::InstOpCode::ldrw, node, tmp1Reg,
```
Comment (knn-k) [inline runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp]: `isLong` is declared nowhere.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp [hunk 4]
```diff
@@ -322,8 +322,174 @@ J9::ARM64::TreeEvaluator::DIVCHKEvaluator(TR::Node *node, TR::CodeGenerator *cg)
...
+   // One of the object is array. Test equality of two objects' classes.
+#ifdef OMR_GC_COMPRESSED_POINTERS
+   // must read only 32 bits
+   generateTrg1MemInstruction(cg, TR::InstOpCode::ldrw, node, tmp2Reg,
+         new (cg->trHeapMemory()) TR::MemoryReference(obj2Reg, (int32_t) TR::Compiler->om.offsetOfObjectVftField(), cg));
+   generateTrg1MemInstruction(cg, TR::InstOpCode::ldrw, node, tmp1Reg,
+         new (cg->trHeapMemory()) TR::MemoryReference(obj1Reg, (int32_t) TR::Compiler->om.offsetOfObjectVftField(), cg));
+#else
+   generateTrg1MemInstruction(cg, isLong ? TR::InstOpCode::ldrx : TR::InstOpCode::ldrw, node, tmp2Reg,
+         new (cg->trHeapMemory()) TR::MemoryReference(obj2Reg, TR::Compiler->om.offsetOfObjectVftField(), TR::Compiler->om.sizeofReferenceAddress(), cg));
```
Comment (knn-k) [inline runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp]: There is no `MemoryReference` constructor available of type `(TR::Register *, int, int, TR::CodeGenerator *)`.

**Comment by aarongraham9** [comment]
Rebased off of `master`, addressed @knn-k's comments. Any thoughts @0xdaryl.

**Comment by 0xdaryl** [comment]
@knn-k , please re-review and let me know if you think it is ready to merge.

**Comment by 0xdaryl** [comment]
Jenkins compile alinux64xl jdk11

**Comment by 0xdaryl** [comment]
@aarongraham9 , please see CI compile failures.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp [hunk 5]
```diff
@@ -323,8 +323,172 @@ J9::ARM64::TreeEvaluator::DIVCHKEvaluator(TR::Node *node, TR::CodeGenerator *cg)
       snippet->gcMap().setGCRegisterMask(0xffffffff);
       }
 
-   cg->evaluate(node->getFirstChild());
+TR::Register *
```
Comment (knn-k) [inline runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp]: The closing `}` of `DIVCHKEvaluator()` is accidentally deleted as the CI Build failure shows.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp (hunk 3 above)
Comment (knn-k) [reply runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp]: You don't need to switch the `ldr` instruction based on `TR::Compiler->target.is64Bit()` in the non-compressed pointer case.  There is no 32-bit target environment for AArch64.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp [hunk 6]
```diff
@@ -323,8 +323,172 @@ J9::ARM64::TreeEvaluator::DIVCHKEvaluator(TR::Node *node, TR::CodeGenerator *cg)
...
+      generateTrg1MemInstruction(cg, TR::InstOpCode::ldrw, node, tmp1Reg,
+            new (cg->trHeapMemory()) TR::MemoryReference(tmp1Reg, offsetof(J9Class, classDepthAndFlags), cg));
+
+      loadConstant32(cg, node, J9AccClassRAMArray, tmp2Reg);
+      generateTrg1Src2Instruction(cg, TR::InstOpCode::andw, node, tmp2Reg, tmp1Reg, tmp2Reg);
+
+      generateCompareImmInstruction(cg, node, tmp2Reg, NULLVALUE);
+
+      snippetLabel = generateLabelSymbol(cg);
+      gcPoint = generateConditionalBranchInstruction(cg, TR::InstOpCode::b_cond, node, snippetLabel, TR::CC_EQ);
```
Comment (knn-k) [inline runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp]: You can use `generateCompareBranchInstruction(TR::InstOpCode::cbzx)` for comparing with zero to branch, replacing `generateCompareImmInstruction()` and `generateConditionalBranchInstruction()`.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp [hunk 7]
```diff
@@ -323,8 +323,172 @@ J9::ARM64::TreeEvaluator::DIVCHKEvaluator(TR::Node *node, TR::CodeGenerator *cg)
...
+
+         // Loading the Class Pointer -> classDepthAndFlags
+#ifdef OMR_GC_COMPRESSED_POINTERS
+         generateTrg1MemInstruction(cg, TR::InstOpCode::ldrw, node, tmp1Reg,
+               new (cg->trHeapMemory()) TR::MemoryReference(obj1Reg, (int32_t) TR::Compiler->om.offsetOfObjectVftField(), cg));
+#else
+         generateTrg1MemInstruction(cg, isLong ? TR::InstOpCode::ldrx : TR::InstOpCode::ldrw, node, tmp1Reg,
+               new (cg->trHeapMemory()) TR::MemoryReference(obj1Reg, TR::Compiler->om.offsetOfObjectVftField(), cg));
+#endif
+         TR::TreeEvaluator::generateVFTMaskInstruction(cg, node, tmp1Reg);
```
Comment (knn-k) [inline runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp]: Why load from VFT field of `obj1` and generate VFTMask again?  You can reuse the value in `tmp1Reg` at `generateCompareInstruction(cg, node, tmp1Reg, tmp2Reg)`.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp [hunk 8]
```diff
@@ -323,8 +323,172 @@ J9::ARM64::TreeEvaluator::DIVCHKEvaluator(TR::Node *node, TR::CodeGenerator *cg)
...
+         generateTrg1MemInstruction(cg, TR::InstOpCode::ldrw, node, tmp1Reg,
+               new (cg->trHeapMemory()) TR::MemoryReference(tmp1Reg, offsetof(J9Class, classDepthAndFlags), cg));
+
+         // We need to perform a X & OBJECT_HEADER_SHAPE_MASK
+         generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::andimmw, node, tmp1Reg, tmp1Reg, OBJECT_HEADER_SHAPE_MASK << J9AccClassRAMShapeShift);
```
Comment (knn-k) [inline runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp]: The instruction encoding of `andimmw` is tricky.  You need to look at `logicBinaryEvaluator()` in OMR.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp [hunk 9]
```diff
@@ -323,8 +323,172 @@ J9::ARM64::TreeEvaluator::DIVCHKEvaluator(TR::Node *node, TR::CodeGenerator *cg)
...
+      if (!node->isArrayChkReferenceArray2())
+         {
+#ifdef OMR_GC_COMPRESSED_POINTERS
+         generateTrg1MemInstruction(cg, TR::InstOpCode::ldrw, node, tmp1Reg,
+               new (cg->trHeapMemory()) TR::MemoryReference(obj2Reg, (int32_t) TR::Compiler->om.offsetOfObjectVftField(), cg));
+#else
+         generateTrg1MemInstruction(cg, isLong ? TR::InstOpCode::ldrx : TR::InstOpCode::ldrw, node, tmp1Reg,
+               new (cg->trHeapMemory()) TR::MemoryReference(obj2Reg, TR::Compiler->om.offsetOfObjectVftField(), cg));
+#endif
+         TR::TreeEvaluator::generateVFTMaskInstruction(cg, node, tmp1Reg);
```
Comment (knn-k) [inline runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp]: I think you can reuse the value in `tmp2Reg` here.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp [hunk 10]
```diff
@@ -323,8 +323,172 @@ J9::ARM64::TreeEvaluator::DIVCHKEvaluator(TR::Node *node, TR::CodeGenerator *cg)
...
+         generateTrg1MemInstruction(cg, TR::InstOpCode::ldrw, node, tmp1Reg,
+               new (cg->trHeapMemory()) TR::MemoryReference(tmp1Reg, offsetof(J9Class, classDepthAndFlags), cg));
+
+         loadConstant32(cg, node, J9AccClassRAMArray, tmp2Reg);
+         generateTrg1Src2Instruction(cg, TR::InstOpCode::andw, node, tmp2Reg, tmp1Reg, tmp2Reg);
+
+         TR::Instruction * i = generateCompareImmInstruction(cg, node, tmp2Reg, NULLVALUE);
```
Comment (knn-k) [inline runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp]: The variable `i` is not used afterwards.
You can use `generateCompareBranchInstruction(TR::InstOpCode::cbzx)` for comparing with zero to branch.

**Comment by aarongraham9** [comment]
Thank you both! I'll address the feedback and push a new version shortly.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp (hunk 8 above)
Comment (knn-k) [reply runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp]: `logicImmediateHelper()` is now available from the outside of OMR BinaryEvaluator.cpp.
You can encode the immediate value for `andimmw` using it.

**Pull request closed without merging**